src/shared/parameter_v2/config/examples/Example.yaml

```diff
@@ -0,0 +1,152 @@
+    max: 5
+    value: 3
+    description: >
+      Can be any integer value in the range [min, max].
+      The minimum value must be >= 0
+- float:
+    name: example_float_param
+    min: 1.1
+    max: 9.01
+    value: 4.04
+    description: >
+      Can be any value in the range [min, max]
+- double:
+    name: example_double_param
+    min: 0
+    max: 5.4
+    value: 0.2
+    description: >
+      Can be any value in the range [min, max]
+
+      Will not be generated for C code, because we assume that is firmware code
+      that cannot represent doubles.
+- string:
+    name: example_string_param
+    value: "Hello World"
+    description: >
+      Any string value
+- enum:
+    name: example_enum_param
+    enum: RefboxGameState
+    value: HALT
+    description: >
+      Accepts enum values from the specified enum. For example if there was an enum defined as
+
+      enum TeamColour {
+        YELLOW,
+        BLUE
+      }
+
+      Then this parameter would only accept TeamColour.YELLOW and TeamColour.BLUE as valid values.
+
+      The target that builds/defines the enum must be included in the enum_deps attribute
+      of the bazel rule (otherwise compilation should fail).
+
+      Will only generate for CPP code.
+- factory:
+    name: example_factory_param
+    index_type: std::string
+    type_to_create: Play
+    value: "HaltPlay"
+    description: >
+      Accepts all registered IndexType values for the GenericFactory<index_type, type_to_create>.
+
+      For example, if there was a GenericFactory<std::string, Play> with 3 entries registered
+      as ("Stop", "Defense", "Offense"), then this parameter would only accept
+      "Stop", "Defense", and "Offense" as valid values.
+
+      Will only generate for CPP code.
+
+#########################################################################################
+# The following are example of special attributes that can be applied to any parameter
+#########################################################################################
+-
+# Make a parameter value optional (CPP only)
+- bool:
+    name: example_optional_parameter
+    value: false
+    optional: true
+    description: >
+      Optional parameters will generate with std::optional as their value return type.
+      They can accept std::nullopt as a value. If optional is not specified, it is
+      assumed to be false.
+
+      Will only generate for CPP code.
+# Make a parameter a constant value
+- int:
+    name: example_const_parameter
+    min: 0
+    max: 255
+    value: 50
+    const: true
+    description: >
+      Parameters with the const attribute do not generate with an API to change their values.
+      They are constants. If const is not specified, it is assumed to be false.
+
+      Only parameters that are marked as const will be generated for C code. We assume
+      that C code is firmware that we cannot change at runtime because it is running
+      on the robots (not the main AI computer).
```

Comment:
What happens when we change values in the yaml? Do we have to flash firmware again? @akhilveeraghanta had an idea about sending a proto to the robots to update dynamic params, so we can avoid flashing firmware just to update dynamic params. In which case, dynamic params don't need to be consts.

Reply:
There still may be the distinction between "constants" and "dynamic params", even for firmware.

ie. we may want most values to be compiled-in, and only allow some to be changed. We don't necessarily want to rely on wifi/radio communication to update some possibly outdated set of params on the robots every time they boot up IMO

For now in the current proposal this will require re-flashing firmware to get updated values.

Reply:
I think that constants are actually extremely useful for performance reasons, as there are some things (for example, robot mass) that are used very often, and we don't want to have to worry about spending the compute to make sure they're constantly updated. Obviously this would only impact a very small handful of cases, but I think it's still a valid consideration.

Reply:
OK so we'd want to track versions of these dynamic param constants in terms of fw hash?

Reply:
Yeah, I think we'd need to generate a hash of all the constants, and then compare that with whatever is in firmware to make sure everything matches up (and warn if not). It's definitely a bit of a pain, so this could very reasonably be a slightly longer term thing.

Reply:
this would also depend how we manage our `firmware_hash` system anyway? ie. to check if the firmware on the robots matches the version running with `ai`. I imagined this may be a hash of the built target rather than all the inputs, which would implicitly contain the firmware constants.

Reply:
Yeah, if those constants were compiled in then you'd be correct.